erlang: cross compile support

IslandUsurper opened the pull request:
###### Description of changes

This, building on top of NixOS/nixpkgs#100835, builds a minimal erlang for the build platform as an input for the cross-compiled erlang. Still need to get the rest of this bootstrap OTP into the path for the rest of the build system to find, as some utilities are not installed by default.
<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

###### Things done

Tested with `nix build .#pkgsCross.aarch64-multiplatform.beam.packages.erlang_25.erlang`. Build completes, but I have no ARM machines to test with.

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

IslandUsurper commented:
There was a review comment about setting an assert so that `enableHipe` isn't set on cross-compiles. I think that'd be too disruptive because it would require everyone doing this to make an overlay or something to override the default value of `enableHipe` on every version of Erlang. The whole purpose of Hipe (as best I can tell) is to speed up compilation when you have are doing a native-arch compile, so it makes sense for it to be disabled automatically.

happysalada commented:
could you ping me again once this is tested ?
Also, just in case if @DianaOlympos has some time to help testing.

DianaOlympos commented:
I sadly do not have an arm machine but i may try to recruit a mac owner

DianaOlympos commented:
@IslandUsurper honestly i would just disable hipe, it is not supported on OTP24+ anymore anyway. And we will drop OTP 23 soon. So let's not deal with it.

IslandUsurper commented:
What's the nixpkgs policy on removing Erlang versions? Just whatever the last 5 versions of Elixir need?

DianaOlympos commented:
@IslandUsurper we only support the last 3 versions modulo 4 versions if the latest OTP was "just released", like right now. So rn, we would support 23, 24, 25 and add 26. And we would probably drop 23 after the next nixos release.

SuperSandro2000 commented on nixos/doc/manual/release-notes/rl-2305.section.md:
```diff
@@ -322,3 +322,5 @@ In addition to numerous new and upgraded packages, this release has the followin
 
 - `k3s` can now be configured with an EnvironmentFile for its systemd service, allowing secrets to be provided without ending up in the Nix Store.
+
+- `beam.interpreters.erlang` may now be cross-compiled using the derivations under `pkgsCross`.
```
Please see the comment under the headline above in the file

SuperSandro2000 commented on pkgs/development/interpreters/erlang/generic-builder.nix:
```diff
@@ -80,7 +83,12 @@ assert javacSupport -> openjdk11 != null;
-
+  isCross = stdenv.hostPlatform != stdenv.buildPlatform;
+  hipeSupport = !isCross && enableHipe;
+  crossPatch = optionalString isCross ''
+    substituteInPlace erts/emulator/Makefile.in \
+      --replace '`utils/find_cross_ycf`' '${pkgsBuildBuild.beam_minimal.interpreters.erlang}/lib/erlang/erts-*/bin/yielding_c_fun'
+  '';
```
```suggestion
```

SuperSandro2000 commented on pkgs/development/interpreters/erlang/generic-builder.nix:
```diff
@@ -107,13 +122,22 @@ stdenv.mkDerivation ({
   enableParallelBuilding = parallelBuild;
 
+  # Clang 4 (rightfully) thinks signed comparisons of pointers with NULL are nonsense
+  prePatch = ''
+    substituteInPlace lib/wx/c_src/wxe_impl.cpp --replace 'temp > NULL' 'temp != NULL'
+
+    ${crossPatch}
+
+    ${prePatch}
+  '';
+
   postPatch = ''
     patchShebangs make
 
     ${postPatch}
   '';
```
```suggestion
  postPatch = ''
    substituteInPlace lib/wx/c_src/wxe_impl.cpp \
      --replace 'temp > NULL' 'temp != NULL'

    patchShebangs make
  '' + optionalString isCross ''
    substituteInPlace erts/emulator/Makefile.in \
      --replace '`utils/find_cross_ycf`' '${pkgsBuildBuild.beam_minimal.interpreters.erlang}/lib/erlang/erts-*/bin/yielding_c_fun'
    '' + postPatch;
```

IslandUsurper commented:
Updated with suggestions.

SuperSandro2000 commented:
Please follow the [contributing guide](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#submitting-changes) when naming your commits.

For this PR, this likely means to just squash everything into one commit.

Are you done with the testing?

jozuas commented:
I have an M3 mac, I'm happy to help with testing anything.

illotum commented:
Building on mac yields:

```
… while evaluating attribute 'buildInputs' of derivation 'erlang-25.3-aarch64-unknown-linux-gnu'
/.../
error: Package ‘libcap-2.67’ in /nix/store/v5qkgmj239gi06priqj1nfildgp4743v-source/pkgs/os-specific/linux/libcap/default.nix:61 is not available on the requested hostPlatform:
         hostPlatform.config = "aarch64-apple-darwin"
/.../
```

LMK If I can help with further testing.

superherointj commented:
ping @IslandUsurper Please fix the merge conflict, and update your PR. Rebasing to master would be ideal.

IslandUsurper committed:
erlang: cross compile support

In order for Erlang to be cross-compiled, it requires a
native-architecture version of Erlang to be in the PATH.
`pkgsBuildBuild.beam_minimal.interperters.erlang` is added to
`depsBuildBuild`.

An attempt to evaluate correctly.

superherointj commented on pkgs/development/interpreters/erlang/generic-builder.nix:
```diff
@@ -113,10 +125,14 @@ stdenv.mkDerivation ({
   postPatch = ''
-    patchShebangs make
+    substituteInPlace lib/wx/c_src/wxe_impl.cpp \
+      --replace 'temp > NULL' 'temp != NULL'
```
`--replace` is deprecated

drupol replied:
Use `--replace-fail` instead.

IslandUsurper replied:
Turns out this whole step isn't necessary when the earliest supported version of Erlang doesn't have this error.

superherointj commented:
Please enable `strictDeps`.

superherointj commented on pkgs/development/interpreters/erlang/generic-builder.nix:
```diff
@@ -113,10 +125,14 @@ stdenv.mkDerivation ({
+    patchShebangs make
+  '' + optionalString isCross ''
+    substituteInPlace erts/emulator/Makefile.in \
+      --replace '`utils/find_cross_ycf`' '${pkgsBuildBuild.beam_minimal.interpreters.erlang}/lib/erlang/erts-*/bin/yielding_c_fun'
```
```suggestion
      --replace '`utils/find_cross_ycf`' '${pkgsBuildBuild.beam_minimal.interpreters.erlang}/lib/erlang/erts-*/bin/yielding_c_fun'
```
`--replace` is deprecated

superherointj commented on pkgs/development/interpreters/erlang/generic-builder.nix:
```diff
@@ -98,9 +103,16 @@ stdenv.mkDerivation ({
-  buildInputs = [ ncurses opensslPackage ]
+  depsBuildBuild = [ perl libxslt ] ++ optional isCross pkgsBuildBuild.beam_minimal.interpreters.erlang;
+
+  buildInputs = [ ncurses opensslPackage zlib ]
```
Why `zlib` wasn't necessary before?

superherointj commented on pkgs/development/interpreters/erlang/generic-builder.nix:
```diff
@@ -113,10 +125,14 @@ stdenv.mkDerivation ({
   postPatch = ''
-    patchShebangs make
+    substituteInPlace lib/wx/c_src/wxe_impl.cpp \
+      --replace 'temp > NULL' 'temp != NULL'
```
```suggestion
      --replace 'temp > NULL' 'temp != NULL'
```
Please use `"` instead of `'`.

drupol replied:
Use `--replace-fail` as well.

superherointj commented on pkgs/development/interpreters/erlang/generic-builder.nix:
```diff
@@ -13,6 +15,7 @@
 , perl
 , runtimeShell
 , autoconf
+, zlib
```
```suggestion
, zlib
```
Would be ideal if alphabetical order was respected.

IslandUsurper replied:
We don't actually need zlib, so it's removed.

superherointj commented on pkgs/development/interpreters/erlang/generic-builder.nix:
```diff
@@ -126,7 +142,7 @@ stdenv.mkDerivation ({
   # built during the build phase of this derivation. Finally, building the
   # documentation requires the erlang-logo.png asset.
-  preConfigure = ''
+  preConfigure = optionals (lib.strings.versionOlder version "22.0") ''
```
```suggestion
  preConfigure = optionals (lib.strings.versionOlder version "22.0") ''
```

There is a problem here:

![image](https://github.com/user-attachments/assets/35fadca3-6336-4d72-a481-92c328b34a5e)

superherointj replied:
Please squash my commit. I've only commited to re-start eval. Such as we know what else is wrong here.

IslandUsurper replied:
nixpkgs doesn't support anything older than 24, so this section is removed.

drupol commented on pkgs/development/interpreters/erlang/generic-builder.nix:
```diff
@@ -98,9 +103,16 @@ stdenv.mkDerivation ({
+    ++ optional hipeSupport gnum4;
 
-  buildInputs = [ ncurses opensslPackage ]
+  depsBuildBuild = [ perl libxslt ] ++ optional isCross pkgsBuildBuild.beam_minimal.interpreters.erlang;
```
I fully agree with this.

superherointj commented:
Errors at `fixupPhase`:

> nix build .#pkgsCross.aarch64-multiplatform.beam.packages.erlang_25.erlang

![image](https://github.com/user-attachments/assets/846459bc-8392-4268-91c5-2f89ed5453a9)

![image](https://github.com/user-attachments/assets/faaa5f53-e677-4cca-909d-bdc54612df3f)

IslandUsurper commented:
I got a build error with `nix build .#pkgsCross.aarch64-multiplatform.beam.packages.erlang_25.erlang` involving `modemmanager`, which comes in from GTK and wxWidgets. But `nix build .#pkgsCross.aarch64-multiplatform.erlang_nox` built successfully.

Going to test removing `zlib`  from inputs. Not sure if it used to be required, and dependencies changed in OTP since I started this PR, or what exactly. I'll also test other versions of OTP, and if all looks good I'll squash down the commits.

IslandUsurper commented:
Oh right. OTP 27 docs are done in exdoc, and we can't cross-compile Elixir yet.

IslandUsurper committed:
elixir: cross compile support

IslandUsurper committed:
rebar: cross compile support

IslandUsurper commented:
That last commit almost certainly could be done better. I thought splicing meant that we didn't have to use `pkgsBuildHost` directly when using inputs in different places.

Still can't build `ex_doc` because something gets messed up in the fixupPhase.

superherointj commented:
The commit messages are not following nixpkgs guidelines.

It should be:
> package-name: xxx

Also, squash the commits per package.

IslandUsurper committed:
rebar3: use erlang from pkgsBuildHost to build

IslandUsurper committed:
ex_doc: use mix from pkgsBuildHost to build

IslandUsurper committed:
hex: use mix from pkgsBuildHost to build

superherointj commented:
> Cross-compile Elixir.
> Cross-compile rebar.

Prefer:

> elixir: fix cross compilation
> rebar: fix cross compilation

Also, lowercase only in all commit titles.

IslandUsurper commented:
```
❯ nix build .#pkgsCross.aarch64-multiplatform.beam_nox.packages.erlang_26.ex_doc
error: builder for '/nix/store/gi6ddzqqnjdl7xx08rr3vw14c9fg0x7c-ex_doc-aarch64-unknown-linux-gnu-0.34.1.drv' failed with exit code 127;
       last 10 log lines:
       > Running phase: fixupPhase
       > shrinking RPATHs of ELF executables and libraries in /nix/store/h2yk4pyqsxj7gqn25icy9w8wcqsjvk35-ex_doc-aarch64-unknown-linux-gnu-0.34.1
       > checking for references to /build/ in /nix/store/h2yk4pyqsxj7gqn25icy9w8wcqsjvk35-ex_doc-aarch64-unknown-linux-gnu-0.34.1...
       > patching script interpreter paths in /nix/store/h2yk4pyqsxj7gqn25icy9w8wcqsjvk35-ex_doc-aarch64-unknown-linux-gnu-0.34.1
       > stripping (with command aarch64-unknown-linux-gnu-strip and flags -S -p) in  /nix/store/h2yk4pyqsxj7gqn25icy9w8wcqsjvk35-ex_doc-aarch64-unknown-linux-gnu-0.34.1/bin
       > removing files for Microsoft Windows
       > wrapping programs in /nix/store/h2yk4pyqsxj7gqn25icy9w8wcqsjvk35-ex_doc-aarch64-unknown-linux-gnu-0.34.1/bin with their runtime deps
       > /nix/store/5vnac2x480gx7abw7kb7w5dk0kxj0a6r-erlang-26.2.5.1-aarch64-unknown-linux-gnu/lib/erlang/bin/../erts-14.2.5.1/bin/dyn_erl: line 1: ELF��@@@8: not found
       > /nix/store/5vnac2x480gx7abw7kb7w5dk0kxj0a6r-erlang-26.2.5.1-aarch64-unknown-linux-gnu/lib/erlang/bin/../erts-14.2.5.1/bin/dyn_erl: line 3: syntax error: unexpected ")"
       > /nix/store/5vnac2x480gx7abw7kb7w5dk0kxj0a6r-erlang-26.2.5.1-aarch64-unknown-linux-gnu/lib/erlang/bin/.erl-wrapped: exec: line 57: ./erts-14.2.5.1/bin/erlexec: not found
       For full logs, run 'nix log /nix/store/gi6ddzqqnjdl7xx08rr3vw14c9fg0x7c-ex_doc-aarch64-unknown-linux-gnu-0.34.1.drv'.
```
Hm. `dyn_erl` seems to have the right header, but that `)` does look odd. I don't know enough about ELF, to really know why it's breaking.

superherointj commented:
@IslandUsurper

I'm no longer active as a maintainer/commiter at nixpkgs.
Please, add other reviewer for this PR (from git history).
Your PR is valid and will be eventually merged. (After build failure for `ex_doc` is sorted.)
Sorry about this.

IslandUsurper commented:
I don't have the permissions to change assignee or reviewers, @superherointj. Guess I'll find a chat room somewhere to bring it up.

superherointj commented:
> I don't have the permissions to change assignee or reviewers, @superherointj. Guess I'll find a chat room somewhere to bring it up.

You can ping people using @.

IslandUsurper commented:
@wegank or @chiroptical, do either of you have time or energy to review this PR?

chiroptical commented:
> @wegank or @chiroptical, do either of you have time or energy to review this PR?

I can try to cross compile to my OSX machine. The code seems fine. I don't have merge privileges so we'll need to find someone to do that assuming this works.

adamcstephens commented:
I tried to build from x86_64-linux to aarch64-linux

```
> nix-build -A pkgsCross.aarch64-multiplatform.beam_minimal.packages.erlang_27
<SNIP>
 GEN    man_index.md
{error,"ERL_TOP environment variable doesn't match the PATH /nix/store/aclb8fi43i91v1ch3vp9ql87zgscqdjy-erlang-27.0/lib/erlang/lib/sasl-4.2.2/ebin/alarm_handler.beam",
       [{otp_man_index,module_path_to_app_vsn,1,
                       [{file,"src/otp_man_index.erl"},{line,80}]},
        {otp_man_index,'-module_table/0-fun-0-',1,
                       [{file,"src/otp_man_index.erl"},{line,55}]},
        {lists,filtermap_1,2,[{file,"lists.erl"},{line,2279}]},
        {otp_man_index,module_table,0,
                       [{file,"src/otp_man_index.erl"},{line,50}]},
        {otp_man_index,gen,1,[{file,"src/otp_man_index.erl"},{line,37}]},
        {init,start_it,1,[]},
        {init,start_em,1,[]},
        {init,do_boot,3,[]}]}{exit,1,[{otp_man_index,gen,1,[{file,"src/otp_man_index.erl"},{line,41}]},{init,start_it,1,[]},{init,start_em,1,[]},{init,do_boot,3,[]}]}
Runtime terminating during boot (1)

Crash dump is being written to: erl_crash.dump...done
make[3]: *** [Makefile:78: man_index.md] Error 1
```

adamcstephens commented:
And for armv7-hf-multiplatform

```
> nix-build -A pkgsCross.armv7l-hf-multiplatform.beam_minimal.packages.erlang_27
<SNIP>
 GEN    html/index.html
qemu-arm: /nix/store/vwa9v14s5h36zmqk976rfzaaa7npkl4g-erlang-27.0-armv7l-unknown-linux-gnueabihf/lib/erlang/bin/escript: Error mapping file: Operation not permitted
make[5]: *** [/build/source/make/doc.mk:62: html/index.html] Error 255
make[5]: Leaving directory '/build/source/lib/stdlib/doc'
make[4]: *** [/build/source/make/otp_release_targets.mk:41: release_docs] Error 2
make[4]: Leaving directory '/build/source/lib/stdlib/doc'
make[3]: *** [/build/source/make/otp_subdir.mk:29: release_docs] Error 2
make[3]: Leaving directory '/build/source/lib/stdlib'
make[2]: *** [/build/source/make/otp_subdir.mk:29: release_docs] Error 2
make[2]: Leaving directory '/build/source/lib'
make[1]: *** [Makefile:439: release_docs] Error 2
make[1]: Leaving directory '/build/source'
make: *** [Makefile:859: install-docs] Error 2
error: builder for '/nix/store/b1afds9xiqrn5jfxakcn3q3y1lhng0mz-erlang-27.0-armv7l-unknown-linux-gnueabihf.drv' failed with exit code 2;
       last 25 log lines:
       > === Entering application stdlib
       > make[4]: Entering directory '/build/source/lib/stdlib/src'
       > make -j24 --jobserver-auth=fifo:/build/GMfifo50229 RELEASE_PATH="/nix/store/vwa9v14s5h36zmqk976rfzaaa7npkl4g-erlang-27.0-armv7l-unknown-linux-gnueabihf/lib/erlang"   release_docs_spec
       > make[5]: Entering directory '/build/source/lib/stdlib/src'
       > make[5]: warning: -j24 forced in submake: resetting jobserver mode.
       > make[5]: Nothing to be done for 'release_docs_spec'.
       > make[5]: Leaving directory '/build/source/lib/stdlib/src'
       > make[4]: Leaving directory '/build/source/lib/stdlib/src'
       > make[4]: Entering directory '/build/source/lib/stdlib/doc'
       > make -j24 --jobserver-auth=fifo:/build/GMfifo50229 RELEASE_PATH="/nix/store/vwa9v14s5h36zmqk976rfzaaa7npkl4g-erlang-27.0-armv7l-unknown-linux-gnueabihf/lib/erlang"   release_docs_spec
       > make[5]: Entering directory '/build/source/lib/stdlib/doc'
       > make[5]: warning: -j24 forced in submake: resetting jobserver mode.
       >  GEN       html/index.html
       > qemu-arm: /nix/store/vwa9v14s5h36zmqk976rfzaaa7npkl4g-erlang-27.0-armv7l-unknown-linux-gnueabihf/lib/erlang/bin/escript: Error mapping file: Operation not permitted
       > make[5]: *** [/build/source/make/doc.mk:62: html/index.html] Error 255
       > make[5]: Leaving directory '/build/source/lib/stdlib/doc'
       > make[4]: *** [/build/source/make/otp_release_targets.mk:41: release_docs] Error 2
       > make[4]: Leaving directory '/build/source/lib/stdlib/doc'
       > make[3]: *** [/build/source/make/otp_subdir.mk:29: release_docs] Error 2
       > make[3]: Leaving directory '/build/source/lib/stdlib'
       > make[2]: *** [/build/source/make/otp_subdir.mk:29: release_docs] Error 2
       > make[2]: Leaving directory '/build/source/lib'
       > make[1]: *** [Makefile:439: release_docs] Error 2
       > make[1]: Leaving directory '/build/source'
       > make: *** [Makefile:859: install-docs] Error 2
       For full logs, run 'nix log /nix/store/b1afds9xiqrn5jfxakcn3q3y1lhng0mz-erlang-27.0-armv7l-unknown-linux-gnueabihf.drv'.
```

adamcstephens commented:
I'm ok with removing docs from the minimal package. We can drop `beam_nodocs` then.

Though we probably want to figure out why ex_doc is failing to cross compile if we can.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/cross-compiling-elixir-project-from-apple-silicon-host-fails-to-use-rebar3-even-in-rosetta-enabled-x86-64-vm/60219/3

onny commented:
please rebase

adamcstephens commented:
May want to wait for https://github.com/NixOS/nixpkgs/pull/448033 or rebase on top of that.

onny reviewed (approved):
I really can't tell if this is the best approach so far but it works for me and code looks good

onny commented:
I get this one error when trying to build it

```
[onny@nixos:~/projects/nixpkgs]$ NIXPKGS_ALLOW_UNFREE=1 nix build --impure -L .#pkgsCross.riscv64.pleroma
error:
       … while calling the 'derivationStrict' builtin
         at «nix-internal»/derivation-internal.nix:37:12:
           36|
           37|   strict = derivationStrict drvAttrs;
             |            ^
           38|

       … while evaluating derivation 'pleroma-riscv64-unknown-linux-gnu-2.10.0'
         whose name attribute is located at /nix/store/h71mdps3z0hg0qpy67jprsbvb0jgdq48-source/pkgs/stdenv/generic/make-derivation.nix:537:13

       … while evaluating attribute 'configurePhase' of derivation 'pleroma-riscv64-unknown-linux-gnu-2.10.0'
         at /nix/store/h71mdps3z0hg0qpy67jprsbvb0jgdq48-source/pkgs/development/beam-modules/mix-release.nix:172:5:
          171|
          172|     configurePhase =
             |     ^
          173|       attrs.configurePhase or ''

       (stack trace truncated; use '--show-trace' to show the full, detailed trace)

       error: undefined variable 'pkgs'
       at /nix/store/h71mdps3z0hg0qpy67jprsbvb0jgdq48-source/pkgs/development/interpreters/elixir/generic-builder.nix:87:27:
           86|     # is executed at build time, not the spliced `inputs.erlang`.
           87|     nativeBuildInputs = [ pkgs.beam_minimal.interpreters.erlang makeWrapper ];
             |                           ^
           88|     buildInputs = [ erlang ];
```

adamcstephens commented:
You can see from the CI checks that there are eval failures. :)

adamcstephens reviewed (commented):
All references to the pkgsBuildBuild erlang need to be updated to access the same major version as the non-build erlang. This is done in a couple places (elixir/erlang generic) but not all. Otherwise the default erlang version will be used for build, which could be newer than the runtime erlang and thus cause problems.

I tried to find my branch from a few months ago that tried to tackle this, but I've apparently lost it. One option would be to compute the build erlang once in `pkgs/development/beam-modules/default.nix` and it'll be in scope for all the children.

> Compiled BEAM Code, NIF Libraries, and Drivers - Compiled code can be loaded on at least two subsequent releases. To achive the highest possible performance for Erlang code, ensure it is compiled using the same release as the one it will be deployed on.
>
> Loading on previous releases is not supported.

https://www.erlang.org/doc/system/misc.html?#compatibility

e.g. right now:
```nix
nix-repl> beam27Packages.rebar3.buildPhase
"HOME=. /nix/store/g0a7iivxzs1yifgiay4bmiy9v1qb3m9s-erlang-28.3.1/bin/escript bootstrap\n"
```
